### ISSUE TITLE
Make TorchIO transforms friendlier to use by related libraries

### ISSUE THREAD

**Issue by fepegar** [fepegar/torchio#221]
**🚀 Feature**

**Motivation**

Other related libraries such as MONAI (https://github.com/Project-MONAI/MONAI/issues/493) and Eisen (https://github.com/eisen-ai/eisen-core/issues/20) are interested in using TorchIO as a dependency. For that, either TorchIO transforms are adapted to accept their data format or their data is adapted to be used with our transforms. Users such as @hjmjohnson have expressed their interest in this connection: https://github.com/Project-MONAI/MONAI/pull/500#issuecomment-640086880

This is a win-win because they won't need to reimplement any transforms and TorchIO will gain visibility.

MONAI's efforts: https://github.com/Project-MONAI/MONAI/pull/500
Eisen's efforts: https://github.com/eisen-ai/eisen-extras/pull/1

**Pitch**

It would be nice to study the way their transforms and data format work and make their life easier by adapting our transforms.

**Alternatives**

Do nothing with TorchIO and offer support for their adaptors.

**Additional context**

Related to #203.

**Comment by neuronflow** [comment]
I post here as my issue seems to be related. I would love to include the promising augmentation techniques of TorchIO in my existing deep learning pipeline. Is there a way I can use them in the
`__getitem__(self, i)` function of my Dataset where I also do my existing augmentation and preprocessing?

Providing TorchIO full governance over my dataset object seems to make numerous adjustments in my code necessary.

**Comment by fepegar** [comment]
Hi @neuronflow. Sure! You can pass 4D PyTorch tensors or NumPy arrays to TorchIO transforms. For example:

```python
import torch
from torchio import RandomElasticDeformation
transform = RandomAffine()
image = torch.rand(1, 256, 256, 178)
transformed = transform(image)
```

Would that work for you?

**Comment by neuronflow** [comment]
Hi @fepegar, thanks for the quick response. In my case I have a t1, t1c, t2, flair image and a label all as numpy arrays. Think of BraTS data. 

From my understanding the spatial transforms you offer should be applied to all images while the intensity images not to the labels.

How can I make sure that the same transform is applied to all of my images?

```
import torch
from torchio import RandomAffine
transform = RandomAffine()
image_t1 = torch.rand(1, 256, 256, 178)
transformed_t1 = transform(image_t1)
image_t2 = torch.rand(1, 256, 256, 178)
transformed_t2 = transform(image_t2)
```
Would that work or do I need to provide a seed and vary this on each iteration?

Also a specific question about RandomAffine(): https://torchio.readthedocs.io/transforms/augmentation.html#torchio.transforms.RandomAffine

If I set `degrees`  to `10` will it always rotate 10 degrees or is this rather the maximum possible rotation?

I especially like your package because it provides "medically plausible" augmentations. Basically I want to apply a whole bunch of augmentations on each iteration through my training set.

**Comment by fepegar** [comment]
> Hi @fepegar, thanks for the quick response. In my case I have a t1, t1c, t2, flair image and a label all as numpy arrays. Think of BraTS data.

Sounds like a job for this library.

> From my understanding the spatial transforms you offer should be applied to all images while the intensity images not to the labels.

Correct.

> How can I make sure that the same transform is applied to all of my images?
> Would that work or do I need to provide a seed and vary this on each iteration?

I think the best you can do is create a `Subject` and extract the NumPy arrays from it:

```python
from torchio import Subject, Image, RandomAffine
subject = Subject(
    t1=ScalarImage(tensor=image_t1),  # this class is new
    t1c=ScalarImage(tensor=image_t1c),
    t2=ScalarImage(tensor=image_t2),
    flair=ScalarImage(tensor=image_flair),
)
transform = RandomAffine()
transformed = transform(subject)
image_t1_transformed = transformed.image_t1.numpy()
# etc.
```

Are your images preprocessed? Wouldn't you like to take the spacing into account for the transformations?

> If I set degrees to 10 will it always rotate 10 degrees or is this rather the maximum possible rotation?

The docs say "If only one value 𝑑 is provided, θ∼U(−𝑑,𝑑)". If you set degrees to 10, the rotations will be between -10 and 10 degrees.

> I especially like your package because it provides "medically plausible" augmentations. Basically I want to apply a whole bunch of augmentations on each iteration through my training set.

Thanks! That's the idea :) Have you checked the tutorial on Google Colab?

**Comment by neuronflow** [comment]
> Are your images preprocessed? Wouldn't you like to take the spacing into account for the transformations?
> 
Yes, my images are preprocessed to isotrope 1mm voxels.

> The docs say "If only one value 𝑑 is provided, θ∼U(−𝑑,𝑑)". If you set degrees to 10, the rotations will be between -10 and 10 degrees.
> 
Sorry, now I do understand the formula :)

> > I especially like your package because it provides "medically plausible" augmentations. Basically I want to apply a whole bunch of augmentations on each iteration through my training set.
> 
> Thanks! That's the idea :) Have you checked the tutorial on Google Colab?

Yes, the google colab was my starting point. however there you create subjects from nifti files and then use subjects to construct a full blown dataset..hence my impression was that I could not use TorchIO in this modular way and stick to a standard pytorch dataset.

**Comment by fepegar** [comment]
> can compose and oneof be stacked within each other?

Yes.

> your example does not deal with labels? how to tell torchio that something is a label when using the above approach?

You can use `torchio.LabelMap`.

**Comment by fepegar** [comment]
Resolved in #252.

**Comment by neuronflow** [comment]
Hi I am trying to marry MONAI with torchio transforms following this tutorial:
https://github.com/Project-MONAI/Tutorials/blob/master/integrate_3rd_party_transforms.ipynb

So far it is working great but I now integrated:
```
        RandomAffine(scales=(0.85, 1.15), degrees=(13), image_interpolation="bspline", p=1, keys=["images", "seg"]),
```
obviously I don't want to interpolate my labels with b-spline but I need to rotate them too. How to solve this? 

I could run the transform twice, once for each with different interpolation. Using seed I could set it up that both should get the same random rotation. But now how do I correctly set a random seed which will be the same for within each iteration of the dataloader but change results over time?

Hope my question becomes clear.

**Comment by romainVala** [comment]
Hi
the interpolation is  nearest neighbor for the label (or actually, all image that are not intensity)
https://github.com/fepegar/torchio/blob/08af12bd138ef3cb0031d103e1c751d187a4ccd5/torchio/transforms/augmentation/spatial/random_affine.py#L169

and the same is done for RandomElasticDeformation

So it should be fine, as soon as you correctly setup label image as torchio.LABEL

**Comment by neuronflow** [comment]
how to achieve that in the MONAI universe? I understand how it works in the torchio world, when giving torchio control over the dataloader etc. but I struggle to wrap my head around it in MONAI :X

**Comment by neuronflow** [comment]
In contrast in MONAI I can supply a tuple with interpolation modes for the respective keys:
```
RandRotated(keys=["image_1", "image_2", "label"], range_x=20.0, range_y=20.0, range_z=20.0,
                     prob=1.0, mode=("bilinear", "bilinear", "nearest")), 
```
I hope my problem becomes clear now.

**Comment by romainVala** [comment]
yes it is clear, but I do not have any experience in MONAI ...
I think the easiest way to go is to construct a torchio subject as posted by fernando
https://github.com/fepegar/torchio/issues/221#issuecomment-658426003
your just need to define your label as a LabelMap

```
subject = Subject(
    t1=ScalarImage(tensor=image_t1),  # this class is new
    label=LabelMap(tensor=label_tensor),
)
```

then when you apply a transform to the subject :
```
transform = RandomAffine()
transformed = transform(subject)
image_t1_transformed = transformed.t1.numpy()  # or transformed['t1'].numpy()
# etc.

```
then nearest will be automatically apply to the label whateve interpolation mode is specify in RandomAffine

I hope it helps

**Comment by neuronflow** [comment]
Thanks for your help, I think I solved it by implementing a custom transform as wrapper!
https://github.com/Project-MONAI/Tutorials/issues/18#issuecomment-690718678

**Comment by fepegar** [comment]
Thanks @romainVala for your help and @neuronflow for reporting and sharing your solution!
